only_otters/scrapetools/hquery.py

```diff
@@ -0,0 +1,317 @@
+# import util
+# import autobrowser
+
+
+class HierarchicalXPathQuery:
+
+    @classmethod
+    def resolve_mode(cls, name, high=False):
+        n = (cls.MODES if not high else cls.HIGHER_ORDER_MODES).get(name)
+        if n is None:
+            raise UserWarning('No such %smode: %s' % ('higher-order ' * high, name))
+        return n
+
+    @classmethod
+    def resolve_pipe(cls, name):
+
+        if ':' in name:
+            ho_pipe, pipename = name.split(':', maxsplit=1)
+
+            hon = cls.HIGHER_ORDER_PIPES.get(ho_pipe)
+            n = cls.PIPES.get(pipename)
+
+            if hon is None:
+                raise UserWarning('No such high-order pipe: %s' % ho_pipe)
+
+            if n is None:
+                raise UserWarning('No such pipe: %s' % pipename)
+
+            return partial(hon, n)
+
+        n = cls.PIPES.get(name)
+        if n is None:
+            raise UserWarning('No such pipe: %s' % name)
+        return n
+
+    @classmethod
+    def resolve_pipe_expr(cls, expr, pipe_prefix=None):
+
+        if pipe_prefix is not None:
+            expr = pipe_prefix + ' ' + expr
+        print(expr)
+
+        regex = r'^\s*\$\s*([a-z]:)?([a-z]+)?\s*{\s*((?:[\w:]+\s*[,]?\s*)+)\s*}'
+        # => $ <modes>{ func, func, func, ... } <query>
+        # see 'query.yml' for an example
+
+        pipes = []
+        modes = []
+        ho_mode = None
+        m = re.match(regex, expr)
+
+        while m is not None:
+            _ho_mode, _modes, _pipes = m.groups()
+
+            # Resolve modes & pipes before runnning Xpath expr
+
+            # Parse modes
+            if _ho_mode is not None:
+                _ho_mode = _ho_mode.strip(':')
+                if len(_ho_mode) > 1 or ho_mode:
+                    raise UserWarning('There can only be one \
+                        mode of higher-order. %s' % ho_mode)
+                ho_mode = cls.resolve_mode(_ho_mode, high=True)
+
+            # Parse pipes
+            _pipes = _pipes.split(',')
+            _pipes = list(map(str.strip, _pipes))
+
+            pipes.extend(_pipes or [])
+            modes.extend(_modes or [])
+
+            expr = expr[m.span()[1]:]
+            m = re.match(regex, expr)
+
+        # Resolve pipes + ho_mode
+        for idx, pipename in enumerate(pipes):
+            fn = cls.resolve_pipe(pipename)
+            if ho_mode is not None:
+                fn = partial(ho_mode, fn)
+            pipes[idx] = fn
+
+        for idx, mode in enumerate(modes):
+            fn = cls.resolve_mode(mode)
+            modes[idx] = fn
+
+        return expr, pipes[::-1], modes, ho_mode
+
+
+
+    @classmethod
+    def resolve_xpath(cls, element, expr, pipe_prefix=None):
+
+        expr, pipes, modes, _ = cls.resolve_pipe_expr(expr, pipe_prefix=pipe_prefix)
+
+        result = element.xpath(expr)
+
+        for pipe in pipes:
+            result = pipe(result)
+
+        for mode in modes:
+            result = mode(result)
+
+        return result
+
+    PIPES = {
```

Comment:
Constants like those should be separated from the actual methods, not sit inbetween them.